test/screens/v2/widget_instance/train_crowding_test.exs

```diff
@@ -5,6 +5,16 @@ defmodule Screens.V2.WidgetInstance.TrainCrowdingTest do
   alias Screens.V2.WidgetInstance.TrainCrowding, as: WidgetInstance
   alias Screens.Vehicles.Vehicle
 
+  defp put_crowding_levels(widget, carriages) do
+    %{
+      widget
+      | prediction: %{
+          widget.prediction
+          | vehicle: %{widget.prediction.vehicle | carriages: carriages}
+        }
+    }
+  end
```

Comment:
There's a cool [convenience function](https://hexdocs.pm/elixir/Kernel.html#put_in/2) (secretely a macro 😈) that makes updates to deeply-nested structures like this easier:

```suggestion
  defp put_crowding_levels(widget, carriages) do
    put_in(widget.prediction.vehicle.carriages, carriages)
  end
```

Just an FYI though; this is fine as-is. There are also some related functions/macros that work in a similar way, like `update_in`, `get_and_update_in`. They're all in the `Kernel` module.

Reply:
Oooo NICE! Good to know!